File: api/src/routes/activities.py
from __future__ import annotations

import json
import uuid

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from src.db import get_db, now_utc
from src.models import ActivityCreate, ActivityResponse, ActivityUpdate
from src.schema import WAYPOINT_ALLOWED_STAGES

router = APIRouter(
    prefix="/api/v1/batches/{batch_id}/activities",
    tags=["activities"],
)


def _serialize_details(details) -> str | None:
    """Serialize a Pydantic details model (or dict) to JSON string."""
    if details is None:
        return None
    if hasattr(details, "model_dump"):
        return json.dumps(details.model_dump())
    return json.dumps(details)


def _deserialize_details(raw: str | None) -> dict | None:
    """Deserialize a JSON string from the DB to a dict."""
    if raw is None:
        return None
    return json.loads(raw)


@router.post("", status_code=201, response_model=ActivityResponse)
async def create_activity(
    batch_id: str,
    body: ActivityCreate,
    request: Request,
):
    db = get_db(request)
    batch = await db.query_one("SELECT * FROM batches WHERE id = ?", (batch_id,))
    if not batch:
        return JSONResponse(
            status_code=404,
            content={
                "error": "not_found",
                "message": "Batch not found",
            },
        )
    if batch["status"] != "active":
        return JSONResponse(
            status_code=409,
            content={
                "error": "conflict",
                "message": "Only active batches can log activities",
            },
        )
    # Validate stage is allowed for current waypoint
    allowed = WAYPOINT_ALLOWED_STAGES.get(batch["stage"], ())
    if body.stage not in allowed:
        return JSONResponse(
            status_code=409,
            content={
                "error": "conflict",
                "message": (
                    f"Stage '{body.stage}' not allowed when batch is at '{batch['stage']}'"
                ),
            },
        )
    activity_id = str(uuid.uuid4())
    now = now_utc()
    details_json = _serialize_details(body.details)
    await db.execute(
        """INSERT INTO activities
           (id, batch_id, stage, type, title,
            details, recorded_at, created_at, updated_at)
           VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)""",
        (
            activity_id,
            batch_id,
            body.stage,
            body.type,
            body.title,
            details_json,
            body.recorded_at,
            now,
            now,
        ),
    )
    row = await db.query_one("SELECT * FROM activities WHERE id = ?", (activity_id,))
    row["details"] = _deserialize_details(row["details"])
    return row


@router.get("", response_model=dict)
async def list_activities(
    batch_id: str,
    request: Request,
    type: str | None = None,
    stage: str | None = None,
    start_time: str | None = None,
    end_time: str | None = None,
):
    db = get_db(request)
    batch = await db.query_one("SELECT id FROM batches WHERE id = ?", (batch_id,))
    if not batch:
        return JSONResponse(
            status_code=404,
            content={
                "error": "not_found",
                "message": "Batch not found",
            },
        )
    sql = "SELECT * FROM activities WHERE batch_id = ?"
    params: list = [batch_id]
    if type:
        sql += " AND type = ?"
        params.append(type)
    if stage:
        sql += " AND stage = ?"
        params.append(stage)
    if start_time:
        sql += " AND recorded_at >= ?"
        params.append(start_time)
    if end_time:
        sql += " AND recorded_at <= ?"
        params.append(end_time)
    sql += " ORDER BY recorded_at DESC"
    rows = await db.query(sql, tuple(params))
    for row in rows:
        row["details"] = _deserialize_details(row["details"])
    return {"items": rows}


@router.patch(
    "/{activity_id}",
    response_model=ActivityResponse,
)
async def update_activity(
    batch_id: str,
    activity_id: str,
    body: ActivityUpdate,
    request: Request,
):
    db = get_db(request)
    row = await db.query_one(
        "SELECT * FROM activities WHERE id = ? AND batch_id = ?",
        (activity_id, batch_id),
    )
    if not row:
        return JSONResponse(
            status_code=404,
            content={
                "error": "not_found",
                "message": "Activity not found",
            },
        )
    updates = body.model_dump(exclude_unset=True)
    if "details" in updates and updates["details"] is not None:
        updates["details"] = _serialize_details(updates["details"])
    if not updates:
        row["details"] = _deserialize_details(row["details"])
        return row
    updates["updated_at"] = now_utc()
    set_clause = ", ".join(f"{k} = ?" for k in updates)
    values = [*updates.values(), activity_id]
    await db.execute(
        f"UPDATE activities SET {set_clause} WHERE id = ?",
        tuple(values),
    )
    row = await db.query_one("SELECT * FROM activities WHERE id = ?", (activity_id,))
    row["details"] = _deserialize_details(row["details"])
    return row


@router.delete("/{activity_id}", status_code=204)
async def delete_activity(
    batch_id: str,
    activity_id: str,
    request: Request,
):
    db = get_db(request)
    row = await db.query_one(
        "SELECT * FROM activities WHERE id = ? AND batch_id = ?",
        (activity_id, batch_id),
    )
    if not row:
        return JSONResponse(
            status_code=404,
            content={
                "error": "not_found",
                "message": "Activity not found",
            },
        )
    await db.execute("DELETE FROM activities WHERE id = ?", (activity_id,))
    return None

File: api/src/routes/activities.ts
import { Hono } from "hono";
import type { AppEnv } from "../app";
import { ActivityCreateSchema, ActivityUpdateSchema } from "../models";
import { WAYPOINT_ALLOWED_STAGES, type BatchStage } from "../schema";
import { notFound, conflict, validationError } from "../lib/errors";
import { nowUtc } from "../lib/time";

const activities = new Hono<AppEnv>();

function isSgMeasurement(type: string, details: Record<string, unknown> | null | undefined): details is Record<string, unknown> {
  return type === "measurement" && !!details && details.metric === "SG" && typeof details.value === "number";
}

activities.post("/", async (c) => {
  const db = c.env.DB;
  const user = c.get("user");
  const batchId = c.req.param("batchId");
  const batch = await db.prepare("SELECT * FROM batches WHERE id = ? AND user_id = ?").bind(batchId, user.id).first<any>();
  if (!batch) return notFound("Batch");
  if (batch.status !== "active") return conflict("Only active batches can log activities");

  const body = await c.req.json().catch(() => null);
  const parsed = ActivityCreateSchema.safeParse(body);
  if (!parsed.success) return validationError(parsed.error.issues);

  const allowed = WAYPOINT_ALLOWED_STAGES[batch.stage as BatchStage] ?? [];
  if (!allowed.includes(parsed.data.stage as any)) {
    return conflict(`Stage '${parsed.data.stage}' not allowed when batch is at '${batch.stage}'`);
  }

  const id = crypto.randomUUID();
  const now = nowUtc();
  const detailsJson = parsed.data.details ? JSON.stringify(parsed.data.details) : null;

  // If this is an SG measurement, also insert a manual reading and link it
  let readingId: string | null = null;
  if (isSgMeasurement(parsed.data.type, parsed.data.details)) {
    readingId = crypto.randomUUID();
    await db
      .prepare(
        `INSERT INTO readings (id, batch_id, device_id, gravity, temperature, battery, rssi, source_timestamp, created_at, source, user_id)
         VALUES (?, ?, 'manual', ?, NULL, NULL, NULL, ?, ?, 'manual', ?)`
      )
      .bind(readingId, batchId, parsed.data.details!.value, parsed.data.recorded_at, now, user.id)
      .run();
  }

  await db
    .prepare(
      `INSERT INTO activities (id, user_id, batch_id, stage, type, title, details, recorded_at, created_at, updated_at, reading_id)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`
    )
    .bind(id, user.id, batchId, parsed.data.stage, parsed.data.type, parsed.data.title,
      detailsJson, parsed.data.recorded_at, now, now, readingId)
    .run();

  const row = await db.prepare("SELECT * FROM activities WHERE id = ?").bind(id).first<any>();
  row.details = row.details ? JSON.parse(row.details) : null;
  return c.json(row, 201);
});

activities.get("/", async (c) => {
  const db = c.env.DB;
  const batchId = c.req.param("batchId");
  const batch = await db.prepare("SELECT id FROM batches WHERE id = ? AND user_id = ?").bind(batchId, c.get("user").id).first();
  if (!batch) return notFound("Batch");

  let sql = "SELECT * FROM activities WHERE batch_id = ?";
  const params: unknown[] = [batchId];
  const type = c.req.query("type");
  const stage = c.req.query("stage");
  const startTime = c.req.query("start_time");
  const endTime = c.req.query("end_time");

  if (type) { sql += " AND type = ?"; params.push(type); }
  if (stage) { sql += " AND stage = ?"; params.push(stage); }
  if (startTime) { sql += " AND recorded_at >= ?"; params.push(startTime); }
  if (endTime) { sql += " AND recorded_at <= ?"; params.push(endTime); }
  sql += " ORDER BY recorded_at DESC";

  const result = await db.prepare(sql).bind(...params).all<any>();
  const items = result.results.map((row: any) => ({
    ...row,
    details: row.details ? JSON.parse(row.details) : null,
  }));
  return c.json({ items });
});

activities.patch("/:activityId", async (c) => {
  const db = c.env.DB;
  const batchId = c.req.param("batchId");
  const activityId = c.req.param("activityId");

  const row = await db.prepare("SELECT * FROM activities WHERE id = ? AND batch_id = ? AND user_id = ?")
    .bind(activityId, batchId, c.get("user").id).first<any>();
  if (!row) return notFound("Activity");

  const body = await c.req.json().catch(() => null);
  const parsed = ActivityUpdateSchema.safeParse(body);
  if (!parsed.success) return validationError(parsed.error.issues);

  const updates: Record<string, unknown> = {};
  if (parsed.data.title !== undefined) updates.title = parsed.data.title;
  if (parsed.data.recorded_at !== undefined) updates.recorded_at = parsed.data.recorded_at;
  if (parsed.data.details !== undefined) {
    updates.details = parsed.data.details ? JSON.stringify(parsed.data.details) : null;
  }

  if (Object.keys(updates).length === 0) {
    row.details = row.details ? JSON.parse(row.details) : null;
    return c.json(row);
  }

  updates.updated_at = nowUtc();
  const setCols = Object.keys(updates).map((k) => `${k} = ?`).join(", ");
  const values = [...Object.values(updates), activityId];
  await db.prepare(`UPDATE activities SET ${setCols} WHERE id = ?`).bind(...values).run();

  // Sync linked reading if SG value or timestamp changed
  if (row.reading_id) {
    const newDetails = parsed.data.details !== undefined
      ? parsed.data.details
      : (row.details ? JSON.parse(row.details) : null);
    if (isSgMeasurement(row.type, newDetails)) {
      const newTimestamp = parsed.data.recorded_at ?? row.recorded_at;
      await db
        .prepare("UPDATE readings SET gravity = ?, source_timestamp = ? WHERE id = ?")
        .bind(newDetails.value, newTimestamp, row.reading_id)
        .run();
    }
  }

  const updated = await db.prepare("SELECT * FROM activities WHERE id = ?").bind(activityId).first<any>();
  updated.details = updated.details ? JSON.parse(updated.details) : null;
  return c.json(updated);
});

activities.delete("/:activityId", async (c) => {
  const db = c.env.DB;
  const batchId = c.req.param("batchId");
  const activityId = c.req.param("activityId");
  const row = await db.prepare("SELECT * FROM activities WHERE id = ? AND batch_id = ? AND user_id = ?")
    .bind(activityId, batchId, c.get("user").id).first<any>();
  if (!row) return notFound("Activity");

  // Delete the linked manual reading if present
  if (row.reading_id) {
    await db.prepare("DELETE FROM readings WHERE id = ?").bind(row.reading_id).run();
  }

  await db.prepare("DELETE FROM activities WHERE id = ?").bind(activityId).run();
  return new Response(null, { status: 204 });
});

export default activities;
